### ISSUE TITLE
[jruby] HTML::DocumentFragment.parse throws a NPE when namespaced attributes are used

### ISSUE THREAD

**Issue by flavorjones** [sparklemotion/nokogiri#1836]
```
# Nokogiri (1.9.0.rc1)
    ---
    warnings: []
    nokogiri: 1.9.0.rc1
    ruby:
      version: 2.5.0
      platform: java
      description: jruby 9.2.4.1 (2.5.0) 2018-11-28 f487d1e OpenJDK 64-Bit Server VM 10.0.2+13-Ubuntu-1ubuntu0.18.04.4
        on 10.0.2+13-Ubuntu-1ubuntu0.18.04.4 [linux-x86_64]
      engine: jruby
      jruby: 9.2.4.1
    xerces: Xerces-J 2.12.0
    nekohtml: NekoHTML 1.9.21
```

```
jruby-9.2.4.1 :001 > Nokogiri::HTML::DocumentFragment.parse "<div xml:base='asdf'>hello</div>"
Traceback (most recent call last):
       16: from org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:72)
       15: from org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:317)
       14: from org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:168)
       13: from org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:199)
       12: from org.jruby.RubyKernel$INVOKER$s$send.call(RubyKernel$INVOKER$s$send.gen)
       11: from org.jruby.RubyKernel.send(RubyKernel.java:2172)
       10: from org.jruby.RubyBasicObject.send(RubyBasicObject.java:1716)
        9: from org.jruby.RubyClass.finvoke(RubyClass.java:502)
        8: from org.jruby.internal.runtime.methods.JavaMethod$JavaMethodZeroOrN.call(JavaMethod.java:846)
        7: from nokogiri.XmlNode$INVOKER$i$0$0$attribute_nodes.call(XmlNode$INVOKER$i$0$0$attribute_nodes.gen)
        6: from nokogiri.XmlNode.attribute_nodes(XmlNode.java:690)
        5: from nokogiri.internals.NokogiriHelpers.getCachedNodeOrCreate(NokogiriHelpers.java:126)
        4: from nokogiri.internals.NokogiriHelpers.constructNode(NokogiriHelpers.java:147)
        3: from nokogiri.XmlNode.setNode(XmlNode.java:583)
        2: from nokogiri.XmlAttr.setNamespaceIfNecessary(XmlAttr.java:106)
        1: from nokogiri.XmlNamespace.createDefaultNamespace(XmlNamespace.java:167)
Java::JavaLang::NullPointerException ()
```

On MRI this is handled OK:

```
2.5.3 :006 > Nokogiri::HTML::DocumentFragment.parse("<div xml:base='asdf'>hello</div>").to_s
 => "<div xml:base=\"asdf\">hello</div>" 
2.5.3 :007 > Nokogiri::HTML::DocumentFragment.parse("<div asdf:qwer='asdf'>hello</div>")
 => #<Nokogiri::HTML::DocumentFragment:0x2b12a98a8590 name="#document-fragment" children=[#<Nokogiri::XML::Element:0x2b12a98a5f84 name="div" attributes=[#<Nokogiri::XML::Attr:0x2b12a98a5908 name="asdf:qwer" value="asdf">] children=[#<Nokogiri::XML::Text:0x2b12a98a4d00 "hello">]>]> 
```

This is happening in the Loofah test suite, which is attempting to sanitize this input. I'd expect that the attribute would be allowed to exist in the document, but honestly any behavior would be preferable to NPE. :-)

**Comment by flavorjones** [comment]
Whoa, mind blown. This throws NPE:

```
jruby-9.2.4.1 :012 > Nokogiri::HTML::DocumentFragment.parse("<div xml:base='asdf'>hello</div>")
Traceback (most recent call last):
       16: from org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:161)
       15: from org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:317)
       14: from org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:168)
       13: from org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:199)
       12: from org.jruby.RubyKernel$INVOKER$s$send.call(RubyKernel$INVOKER$s$send.gen)
       11: from org.jruby.RubyKernel.send(RubyKernel.java:2172)
       10: from org.jruby.RubyBasicObject.send(RubyBasicObject.java:1716)
        9: from org.jruby.RubyClass.finvoke(RubyClass.java:502)
        8: from org.jruby.internal.runtime.methods.JavaMethod$JavaMethodZeroOrN.call(JavaMethod.java:846)
        7: from nokogiri.XmlNode$INVOKER$i$0$0$attribute_nodes.call(XmlNode$INVOKER$i$0$0$attribute_nodes.gen)
        6: from nokogiri.XmlNode.attribute_nodes(XmlNode.java:690)
        5: from nokogiri.internals.NokogiriHelpers.getCachedNodeOrCreate(NokogiriHelpers.java:126)
        4: from nokogiri.internals.NokogiriHelpers.constructNode(NokogiriHelpers.java:147)
        3: from nokogiri.XmlNode.setNode(XmlNode.java:583)
        2: from nokogiri.XmlAttr.setNamespaceIfNecessary(XmlAttr.java:106)
        1: from nokogiri.XmlNamespace.createDefaultNamespace(XmlNamespace.java:167)
Java::JavaLang::NullPointerException ()
```

and this does not:

```
jruby-9.2.4.1 :011 > Nokogiri::HTML::DocumentFragment.parse("<div xml:base='asdf'>hello</div>").to_s
 => "<div xml:base=\"asdf\">hello</div>" 
```

**Comment by headius** [comment]
It appears some logic is allowing a null value to get into a Ruby object where it should not be. The NPE is actually from the inspect call in IRB. Here's similar output from `p` at a command line:

```
$ jruby -v -rnokogiri -e 'p Nokogiri::HTML::DocumentFragment.parse("<div xml:base='asdf'>hello</div>")'
jruby 9.2.10.0-SNAPSHOT (2.5.7) 2020-01-14 1a2ed80 OpenJDK 64-Bit Server VM 25.222-b10 on 1.8.0_222-b10 +jit [darwin-x86_64]
Unhandled Java exception: java.lang.NullPointerException
java.lang.NullPointerException: null
   createDefaultNamespace at nokogiri/XmlNamespace.java:167
  setNamespaceIfNecessary at nokogiri/XmlAttr.java:106
                  setNode at nokogiri/XmlNode.java:583
            constructNode at nokogiri/internals/NokogiriHelpers.java:147
    getCachedNodeOrCreate at nokogiri/internals/NokogiriHelpers.java:126
          attribute_nodes at nokogiri/XmlNode.java:690
                     call at nokogiri/XmlNode$INVOKER$i$0$0$attribute_nodes.gen:-1
                     call at org/jruby/internal/runtime/methods/JavaMethod.java:864
                  finvoke at org/jruby/RubyClass.java:493
                     send at org/jruby/RubyBasicObject.java:1721
                     send at org/jruby/RubyKernel.java:2188
                     call at org/jruby/RubyKernel$INVOKER$s$send.gen:-1
                     call at org/jruby/internal/runtime/methods/JavaMethod.java:417
                     call at org/jruby/runtime/callsite/CachingCallSite.java:172
              processCall at org/jruby/ir/interpreter/InterpreterEngine.java:316
```

**Comment by headius** [comment]
This appears to have been introduced in 1.10.5:

```
[] ~/projects/jruby $ gem uninstall nokogiri

Select gem to uninstall:
 1. nokogiri-1.10.4-java
 2. nokogiri-1.10.5-java
 3. All versions
> 2
Successfully uninstalled nokogiri-1.10.5-java

[] ~/projects/jruby $ jruby -v -Xjit.threshold=0 -rnokogiri -e 'p Nokogiri::HTML::DocumentFragment.parse("<div xml:base='asdf'>hello</div>")'
jruby 9.2.10.0-SNAPSHOT (2.5.7) 2020-01-14 1a2ed80 OpenJDK 64-Bit Server VM 25.222-b10 on 1.8.0_222-b10 +jit [darwin-x86_64]
#<Nokogiri::HTML::DocumentFragment:0x7d6 name="#document-fragment" children=[#<Nokogiri::XML::Element:0x7d4 name="div" attributes=[#<Nokogiri::XML::Attr:0x7d0 name="xml:base" value="asdf">] children=[#<Nokogiri::XML::Text:0x7d2 "hello">]>]>
```

The line in 1.10.7 appears to be here: https://github.com/sparklemotion/nokogiri/blob/a2cea74e2bdad0c0106d7848d093c8faeaf4fd32/ext/java/nokogiri/XmlNamespace.java#L167

Step debugging into that logic, it appears the XmlDocument has a null nsCache field, which causes an NPE here (and would cause another one later in the method). I tried adding `initializeNamespaceCacheIfNecessary` in the XmlDocument constructor that did not call it, but that did not seem to fix the problem.

The document in question is an HtmlDocument, but both of its constructors pass through XmlDocument constructors that now initialize the cache.

**Comment by headius** [comment]
There are **two** constructors in XmlDocument that do not initialize the namespace cache:

Here:

https://github.com/sparklemotion/nokogiri/blob/280a5370072e7995e2c2cbb7e634ed8019a26949/ext/java/nokogiri/XmlDocument.java#L105-L107

And here:

https://github.com/sparklemotion/nokogiri/blob/280a5370072e7995e2c2cbb7e634ed8019a26949/ext/java/nokogiri/XmlDocument.java#L175-L181

The first one never initializes the cache, requiring consumers to make sure they request it be initialized. The second one gets its cache from the incoming document and assumes it's non-null.

I am assuming this field was made lazy to avoid creating the cache for documents that don't do any namespace lookups, but it is not consistently initialized. I'm not going to try to fix that, though, and the following patch appears to resolve this issue:

```diff
diff --git a/ext/java/nokogiri/XmlNamespace.java b/ext/java/nokogiri/XmlNamespace.java
index ed97519e..ac8a985c 100644
--- a/ext/java/nokogiri/XmlNamespace.java
+++ b/ext/java/nokogiri/XmlNamespace.java
@@ -164,6 +164,7 @@ public class XmlNamespace extends RubyObject {
         Document document = owner.getOwnerDocument();
         // check namespace cache
         XmlDocument xmlDocument = (XmlDocument)getCachedNodeOrCreate(runtime, document);
+        xmlDocument.initializeNamespaceCacheIfNecessary();
         XmlNamespace xmlNamespace = xmlDocument.getNamespaceCache().get(prefixValue, hrefValue);
         if (xmlNamespace != null) return xmlNamespace;
```

Alternatively, it could be forced to initialize in the remaining non-initializing constructor, but it will no longer be lazy (unless someone explicitly sets it to null later on):

```diff
diff --git a/ext/java/nokogiri/XmlDocument.java b/ext/java/nokogiri/XmlDocument.java
index 152d2e9e..180b3dcd 100644
--- a/ext/java/nokogiri/XmlDocument.java
+++ b/ext/java/nokogiri/XmlDocument.java
@@ -104,6 +104,7 @@ public class XmlDocument extends XmlNode {
 
     public XmlDocument(Ruby ruby, RubyClass klazz) {
         super(ruby, klazz, createNewDocument());
+        initializeNamespaceCacheIfNecessary();
     }
 
     public XmlDocument(Ruby ruby, Document document) {
```

**Comment by flavorjones** [comment]
I want to note that I opened this bug report against Nokogiri 1.9.0.rc1 which means that the history is more complex than "introduced in 1.10.5," though I'll admit I have largely ceded ownership of the Java codebase to John and Karol (cced above) and they probably have more context than I do.

**Comment by headius** [comment]
In any case something that merged into Nokogiri between 1.10.4 and 1.10.5 caused this regression, because the former appears to work correctly.

**Comment by kares** [comment]
oh right, my bad ... that NS cache usage was a real pain.
might have broke a thingy changing `initializeNamespaceCacheIfNecessary()` despite tests :green_apple: 
was trying to avoid creating too many objects for empty xml documents (which was used extensively)

**Comment by jvshahid** [comment]
I'm having trouble reproducing the issue using the methods in https://github.com/sparklemotion/nokogiri/issues/1836#issuecomment-445711214 and https://github.com/sparklemotion/nokogiri/issues/1836#issuecomment-574244756. In both cases, I am not getting any exceptions. For example:

```
$ bundle exec irb -Ilib -rnokogiri
jruby-9.2.6.0 :001 > Nokogiri::HTML::DocumentFragment.parse("<div xml:base='asdf'>hello</div>")
 => #<Nokogiri::HTML::DocumentFragment:0x7d8 name="#document-fragment" children=[#<Nokogiri::XML::Element:0x7d6 name="div" attributes=[#<Nokogiri::XML::Attr:0x7d2 name="xml:base" value="asdf">] children=[#<Nokogiri::XML::Text:0x7d4 "hello">]>]> 
jruby-9.2.6.0 :002 > 
```

```
$ bundle exec jruby -Ilib -v -rnokogiri -e 'p Nokogiri::HTML::DocumentFragment.parse("<div xml:base='asdf'>hello</div>")'
/home/jvshahid/.rvm/rubies/jruby-9.2.6.0/lib/ruby/stdlib/rubygems/defaults/jruby.rb:34: warning: constant Gem::ConfigMap is deprecated
jruby 9.2.6.0 (2.5.3) 2019-02-11 15ba00b OpenJDK 64-Bit Server VM 10.0.2+13 on 10.0.2+13 +jit [linux-x86_64]
/home/jvshahid/.rvm/rubies/jruby-9.2.6.0/lib/ruby/stdlib/rubygems/defaults/jruby.rb:34: warning: constant Gem::ConfigMap is deprecated
#<Nokogiri::HTML::DocumentFragment:0x7d6 name="#document-fragment" children=[#<Nokogiri::XML::Element:0x7d4 name="div" attributes=[#<Nokogiri::XML::Attr:0x7d0 name="xml:base" value="asdf">] children=[#<Nokogiri::XML::Text:0x7d2 "hello">]>]>
```

**Comment by jvshahid** [comment]
> oh right, my bad ... that NS cache usage was a real pain.

Have we looked into removing the cache?  It was introduced in 2009 in https://github.com/sparklemotion/nokogiri/commit/5526508cc95816ee05cef6f3d5315abdc5f28c45. I am wondering if it is still relevant.

**Comment by kares** [comment]
not really, seemed necessary. there's some other path that depend on the namespace-cache code.

**Comment by flavorjones** [comment]
@jvshahid and @kares - is there any chance you'll get a chance to look at this more deeply in the next few weeks? If not, that's OK, but then I'll remove it from the v1.11.0 milestone -- just want to understand expectations.

**Comment by jvshahid** [comment]
I still can't reproduce this issue locally using commit d03e956d:shrug:. Can you reproduce this in CI? I could try using fly execute to understand the issue better.

**Comment by flavorjones** [comment]
After a little git-bisecting, this seems to have been fixed by @kares in 3dd0d83a27c52b142b6b84596083d04e1dabed8e as part of #1934. This change landed in v1.11.0.
